src/StateMachine.jl

```diff
@@ -240,5 +221,105 @@ function pretty(
         end
         long && print(io, "   ")
         print(io, " ELSE: $(name(next[2], id))")
+    elseif next isa Tuple{}
+        if long
+            print(io, "\n   ")
+            print(io, " END")
+        end
+    else
+        error(" UNKNOWN ")
+    end
+end
+
+# We merge states with identical behavior, bottom-up, to minimize the size of
+# the state machine.  We define `hash` and `==` to take account of only what matters.
+# Specifically, we ignore the `cases::ImmutableVector{CasePartialResult}` of `CodePoint`.
+mutable struct DeduplicatedCodePoint
+    # A label to produce in the code at entry to the code where
+    # this state is implemented, if one is needed.  This is not produced
+    # when this struct is created, but later during code generation.
+    label::Union{Nothing, Symbol}
+
+    # The selected action to take from this state: either
+    # - Case whose tests have all passed, or
+    # - A bound pattern to perform and then move on to the next state, or
+    # - An Expr to insert into the code when all else is exhausted
+    #   (which throws MatchFailure)
+    @_const action::Union{CasePartialResult, BoundPattern, Expr}
+
+    # The next code point(s):
+    # - Tuple{} if the action is a case which was matched or a MatchFailure
+    # - Tuple{DeduplicatedCodePoint} if the action was a fetch pattern. It designates
+    #   the code to perform after the fetch.
+    # - Tuple{DeduplicatedCodePoint, DeduplicatedCodePoint} if the action is a test.  These are the states
+    #   to go to if the result of the test is true ([1]) or false ([2]).
+    @_const next::Union{Tuple{}, Tuple{DeduplicatedCodePoint}, Tuple{DeduplicatedCodePoint, DeduplicatedCodePoint}}
+
+    @_const _cached_hash::UInt64
+    function DeduplicatedCodePoint(action, next)
+        action isa CasePartialResult && @assert action.pattern isa BoundTruePattern
+        new(nothing, action, next, hash((action, next)))
+    end
+end
+Base.hash(code::DeduplicatedCodePoint, h::UInt64) = hash(code._cached_hash, h)
+Base.hash(code::DeduplicatedCodePoint) = code._cached_hash
+function Base.:(==)(a::DeduplicatedCodePoint, b::DeduplicatedCodePoint)
+    a === b ||
+        a._cached_hash == b._cached_hash &&
+        isequal(a.action, b.action) &&
+        isequal(a.next, b.next)
+end
+function ensure_label!(code::DeduplicatedCodePoint, state::BinderState)
+    if code.label isa Nothing
+        code.label = gensym("label", state)
+    end
+end
+
+#
+# Deduplicate a code point, given the deduplications of the downstream code points.
+#
+function dedup(
+    dict::Dict{DeduplicatedCodePoint, DeduplicatedCodePoint},
+    code::CodePoint,
+    state::BinderState)
+    next = if code.next isa Tuple{}
```

Comment:
I wonder if it would make sense to alias these for readability since they are used in a few places and explained at least once.

something like
```
const TerminalNext = Tuple{}
const FallthroughNext = Tuple{<:AbstractCodePoint}
const BranchNext = Tuple{T, T} where T <: AbstractCodePoint
```
(caveat emptor: untested code)

Reply:
I've been bitten by aliases recently.  I could consider this as a separate refactoring.

Reply:
Yeah, aliases sometimes are too lightweight. This might be a case where it's worth creating structs for in a future PR

something like
```
abstract type NextNode end

struct Terminal <: NextNode end

struct FallThrough{T} <: NextNode where {T}
   node::T
end

struct Branch{T} <: NextNode where {T}
   true_branch::T
   false_branch::T
end
```

Reply:
https://github.com/JuliaLang/julia/issues/50241